refactor: RHICOMPL-1355 presentational TabbedRules from PolicyMultiversionRules

vkrizan opened the pull request:
Created presentational component `TabbedRules` out of `PolicyMultiversionRules`

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [X] Access Control
- [X] Cryptographic Practices
- [X] Error Handling and Logging
- [X] Data Protection
- [X] Communication Security
- [X] System Configuration
- [X] Database Security
- [X] File Management
- [X] Memory Management
- [X] General Coding Practices

bastilian commented on src/SmartComponents/PolicyDetails/PolicyMultiversionRules.js:
```diff
@@ -1,38 +1,25 @@
+import { Alert, Label, TabTitleText, PageSection, PageSectionVariants } from '@patternfly/react-core';
+import { TabbedRules } from 'PresentationalComponents';
+
+const tabTitle = ({ ssgVersion, rules }) => (
```
This would probably be better to be a real Component in `TabbedRules`

vkrizan replied:
Can you please give an example with a callback?

bastilian commented on src/PresentationalComponents/TabbedRules/TabbedRules.js:
```diff
@@ -0,0 +1,46 @@
+            profilesAndRules?.map(({ profile, rules }) => (
+                <Tab
+                    key={ 'multiversion-rules-' + profile.id }
+                    title={ tabTitle(profile) }
```
Related to the other comment. This should be something like 

```js 
title={ <TabTitle ssgVersion={ profile.ssgVersion } rules={ profile.rules.length } /> } 
```

with a component above like:

```js
const TabTitle = ({ ssgVersion, rules }) => (
    <TabTitleText>
        <span>SSG { ssgVersion + ' ' }</span>
        <Label color="blue">{ rules  }</Label>
    </TabTitleText>
);
```

vkrizan replied:
That would hardcore the behavior of tab title. Details page shows SSG versions (now), modal would show OS versions. Details page shows number of rules (now), modal would show number of assigned or selected hosts.

I'm targeting to use the callback to construct the title with data that are not directly provided for the `TabbedRules` component. For example number of hosts wouldn't be passed, but the title can still be drawn with it.

It might be changed to some common interface:
```
title={ <tabTitleComponent  profile={profile} /> }
```

bastilian replied:
In both (all multi OS version rules table) instances the title should be the same and show the RHEL version and assigned hosts, we do not need to preserve the rules count and and SSG version, since it is actually a bug.

vkrizan replied:
Once that happens we can update it. At the moment lets do iterative changes that make the presentational component more robust/open.

As mentioned earlier, with callback it is possible to inject other values that are within the scope of a (anonymous) function.

bastilian replied:
Using a "callback" like that is not a great approach. Passing in the a rendered component rather than a function to be called to render, would be better.

vkrizan replied:
I agree, but cannot think of something better. When this would be utilized within the modals its usability might be more clear or some other approach might arise.

vkrizan replied:
Other way could be to provide titles within the array of data. Current `profilesAndRules` would contain array of `{profile, title, rules}`.

bastilian replied:
Actually, we only need to cover the OS tab, because the SSG versions are behind a feature flag still.

So it should not be necessary at all to provide a function or component from the outside at all.

bastilian commented on src/SmartComponents/PolicyDetails/PolicyMultiversionRules.js:
```diff
@@ -1,38 +1,31 @@
+            <TabbedRules
+                profilesAndRules={ mapProfilesWithRules(profiles) }
+                tabTitle={ tabTitle }
+                defaultProfileId={ profiles[0].id }
```
Could the first profile id not be the default within the `TabbedRules` component and only require setting it when necessary and the first one should not be the default opened?

vkrizan replied:
This does not change the opening behavior. The Policy details page does open first tab.

bastilian replied:
What I mean is do we really need to provide a `defaultProfileId`, can we not just set the first item in `profilesAndRules` as the `defaultTab` for the `RoutedTabs` component within `TabbedRules`?

vkrizan replied:
Oh, right! We can do have a default for default. :slightly_smiling_face:

bastilian commented on src/PresentationalComponents/TabbedRules/TabbedRules.js:
```diff
@@ -0,0 +1,46 @@
+} from '@redhat-cloud-services/frontend-components-inventory-compliance/SystemRulesTable';
+import { RoutedTabs } from 'PresentationalComponents';
+
+const TabbedRules = ({ profilesAndRules, tabTitle, defaultProfileId, columns, level, handleSelect, ...rulesTableProps }) => (
```
Default props can also be set by just assigning here like:

```js
{ profilesAndRules = [], columns=['Name'], level=1 ...
```

vkrizan replied:
It's already long enough. Do you know how to format it to several lines?

bastilian replied:
Press the enter key at convenient locations? :D

vkrizan replied:
It looks terrible. Sticking with `defaultProps` this time.

bastilian replied:
It doesn't look too terrible: 

```js
const TabbedRules = ({
    columns = ['Name', 'Severity', 'Ansible'],
    defaultProfileId,
    handleSelect,
    level = 0,
    profilesAndRules,
    tabTitle,
    ...rulesTableProps
}) => {
    // ...
};
```

It also lifts up the defaults in the file and can be seen right away.

akofink commented on src/PresentationalComponents/TabbedRules/TabbedRules.js:
```diff
@@ -0,0 +1,46 @@
+import { RoutedTabs } from 'PresentationalComponents';
+
+const TabbedRules = ({ profilesAndRules, tabTitle, defaultProfileId, columns, level, handleSelect, ...rulesTableProps }) => (
+    <RoutedTabs level={ level } defaultTab={ defaultProfileId }>
```
This `level` used to always be 1 before this. I guess it will be 1 in the policy details -> rules view, but it will be 0 in the edit policy modal. Is that correct?

vkrizan replied:
Yes, level 1 in details and level 0 in create/edit modal. 

Disabling the routing isn't possible with this component. If that would be needed, we can split it to two components, one with routing one without routing. That's the easiest approach.

akofink replied:
I think the routing is pretty useful so far. It'd be great to be able to link users directly to the rule set for editing or viewing. At least in this use case, I don't see us needing tabs without routing.

akofink commented on src/PresentationalComponents/TabbedRules/TabbedRules.js:
```diff
@@ -0,0 +1,46 @@
+    </RoutedTabs>
+);
+
+TabbedRules.propTypes = {
+    profilesAndRules: propTypes.arrayOf(propTypes.object),
+    tabTitle: propTypes.func.isRequired,
+    defaultProfileId: propTypes.string,
+    columns: propTypes.arrayOf(propTypes.string),
+    level: propTypes.number,
+    handleSelect: propTypes.func
+};
```
So, the idea of this new component is to use it for policy details, edit policy modal, and create policy wizard, right? Are you planning to add some prop like `allowSelection` or similar to toggle between the rule display vs. rule selection behavior?

vkrizan replied:
Yes that's the idea. You can pass `tailoringEnabled` boolean. It's redirected to the SystemsRulesTable.

vkrizan replied:
Sorry, `tailoringEnabled` is not even used, instead the `handleSelect` should be set.

akofink reviewed (approved):
Thanks, @vkrizan! I only have a couple small questions, but it seems to continue to work as expected, and I think I understand the intended uses for this new component. I'm interested in your vision, though, for how we'd use it (additional props, etc.) in edit/create policy views.

vkrizan commented:
Okay, I've cleaned it up a bit. Changed the title callback to a node provided through `tabsData` and set auto defaulting to first item. Event keys have `rules-` prepended to avoid possible conflicts.

Would you guys please have one last peak?

akofink reviewed (approved):
Seems to continue working as expected. Thanks!

vkrizan commented on src/SmartComponents/PolicyDetails/PolicyMultiversionRules.js:
```diff
@@ -1,38 +1,37 @@
+import { Alert, Label, TabTitleText, PageSection, PageSectionVariants } from '@patternfly/react-core';
+import { TabbedRules } from 'PresentationalComponents';
+
+const SSGTabTitle = ({ profile: { ssgVersion, rules = [] } }) => (
+    <TabTitleText>
+        <span>SSG { ssgVersion + ' ' }</span>
+        <Label color="blue">{ rules.length }</Label>
+    </TabTitleText>
+);
```
Sure that sounds doable. It would likely be flat, like `SSGRulesTabTitle`. This one would eventually go away. We can definitely use this suggestion for the OS tab that'd be used thought the app.
